### PR TITLE
Network Log Changes

### PR DESCRIPTION
**Pull request by prashelke**

### CODE REVIEW

File: Ginger/GingerCoreNET/ActionsLib/UI/Web/ActBrowserElement.cs [hunk 1]
```diff
@@ -350,5 +383,25 @@ public override bool SerializationError(SerializationErrorType errorType, string
             return false;
         }
 
+        public enum eMonitorUrl
+        {
+            [EnumValueDescription("All Url's")]
```
Comment (Maheshkale447) [inline Ginger/GingerCoreNET/ActionsLib/UI/Web/ActBrowserElement.cs:388]: change Url's to URLs

File: Ginger/GingerCoreNET/ActionsLib/UI/Web/ActBrowserElement.cs [hunk 2]
```diff
@@ -350,5 +383,25 @@ public override bool SerializationError(SerializationErrorType errorType, string
...
+            AllUrl,
+            [EnumValueDescription("Selected Url's")]
```
Comment (Maheshkale447) [inline Ginger/GingerCoreNET/ActionsLib/UI/Web/ActBrowserElement.cs:390]: change Url's to URLs

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs [hunk 3]
```diff
@@ -68,6 +69,12 @@ public class SeleniumDriver : DriverBase, IVirtualDriver, IWindowExplorer, IVisu
     {
         protected IDevToolsSession Session;
         DevToolsSession devToolsSession;
+        Dictionary<string, object> networklogdict;
```
Comment (Maheshkale447) [inline Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs:72]: change name to networkResponseLogDict and networkRequestLogDict

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs [hunk 4]
```diff
@@ -6753,50 +6760,64 @@ public void ActBrowserElementHandler(ActBrowserElement act)
...
+                        
+                    break;
+                case ActBrowserElement.eControlAction.GetNetworkLog:
+                    {
+                        GetNetworkLogAsync(Driver, act).GetAwaiter().GetResult();
+                    }
+                    break;
+                case ActBrowserElement.eControlAction.StopMonitoringNetworkLog:
+                    {
+                        StopMonitoringNetworkLog(Driver, act).GetAwaiter().GetResult();
```
Comment (Maheshkale447) [inline Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs:6817]: Please add check for browser type

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs [hunk 5]
```diff
@@ -8634,115 +8655,222 @@ private void SetUPDevTools(IWebDriver webDriver)
...
+                act.ExInfo = "Action is skipped," + ActBrowserElement.eControlAction.StartMonitoringNetworkLog.ToString() + " Action is not started";
+                act.Status = Amdocs.Ginger.CoreNET.Execution.eRunStatus.Skipped;
+
+            }
+
+        }
+
+        private void OnNetworkRequestSent(object sender, NetworkRequestSentEventArgs e)
+        {
+            if (mAct.GetOrCreateInputParam(nameof(ActBrowserElement.eMonitorUrl)).Value == ActBrowserElement.eMonitorUrl.SelectedUrl.ToString() && mAct.UpdateOperationInputValues != null && mAct.UpdateOperationInputValues.Any(x => e.RequestUrl == x.Param.ToString()))
```
Comment (Maheshkale447) [inline Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs:8746]: add tolower while comparing url

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs [hunk 6]
```diff
@@ -8634,115 +8655,222 @@ private void SetUPDevTools(IWebDriver webDriver)
...
+            {
+                act.AddOrUpdateReturnParamActual("Raw Request", Newtonsoft.Json.JsonConvert.SerializeObject(networklogdictRequest.ToList()));
+                foreach (var val in networklogdictRequest.ToList())
+                {
+                    act.AddOrUpdateReturnParamActual(act.ControlAction.ToString() + val.Key.ToString(), Convert.ToString(val.Value));
+                }
+                act.AddOrUpdateReturnParamActual("Raw Response", Newtonsoft.Json.JsonConvert.SerializeObject(networklogdict.ToList()));
+                foreach (var val in networklogdict.ToList())
+                {
+                    act.AddOrUpdateReturnParamActual(act.ControlAction.ToString() + val.Key.ToString(), Convert.ToString(val.Value));
```
Comment (Maheshkale447) [inline Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs:8674]: add space between concatenating 2 strings

**Review by Maheshkale447 — CHANGES_REQUESTED** [review]

File: Ginger/GingerCoreNET/ActionsLib/UI/Web/ActBrowserElement.cs (hunk 1 above)
Comment (prashelke) [reply Ginger/GingerCoreNET/ActionsLib/UI/Web/ActBrowserElement.cs:388]: Changes Done

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs (hunk 3 above)
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs:72]: Changes Done

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs (hunk 4 above)
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs:6817]: Changes Done

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs [hunk 7]
```diff
@@ -8634,115 +8655,222 @@ private void SetUPDevTools(IWebDriver webDriver)
...
-            //domains.Console.Enable();
-            //domains.Console.ClearMessages();
-            //domains.Console.MessageAdded += messageAdded;
-            //var ff = DevToolsSessionLogLevel.;
-
-            //object p = webDriver.e("console.log('BELLATRIX is cool');");
-            INetwork interceptor = webDriver.Manage().Network;
+        public async Task GetNetworkLogAsync(IWebDriver webDriver,ActBrowserElement act)
+        {
+            if (isStartMonitor)
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs:8664]: Changes Done

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs (hunk 5 above)
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs:8746]: Changes Done

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs (hunk 6 above)
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs:8674]: Changes Done

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs [hunk 8]
```diff
@@ -8634,115 +8655,222 @@ private void SetUPDevTools(IWebDriver webDriver)
...
+        }
+
+        private void OnNetworkRequestSent(object sender, NetworkRequestSentEventArgs e)
+        {
+            if (mAct.GetOrCreateInputParam(nameof(ActBrowserElement.eMonitorUrl)).Value == ActBrowserElement.eMonitorUrl.SelectedUrl.ToString() && mAct.UpdateOperationInputValues != null && mAct.UpdateOperationInputValues.Any(x => e.RequestUrl == x.Param.ToString()))
+            {
+                if(networklogdictRequest.Count > 0)
                 {
-                    builder.AppendLine("<JavaScript content>");
+                    if (networklogdictRequest.ContainsKey(e.RequestUrl.ToString()))
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs:8750]: Changes Done

**Review by Maheshkale447 — APPROVED** [review]

**Pull request merged**